core/storage/btree.rs

```diff
@@ -501,55 +527,68 @@ impl BTreeCursor {
-        };
-
-        let (cell_idx, page_type) = {
-            let page = RefCell::borrow(&page_ref);
-            if page.is_locked() {
-                return Ok(CursorResult::IO);
-            }
+        let state = &self.write_info.state;
+        match state {
+            WriteState::Start => {
+                let page_ref = self.get_current_page()?;
+                let int_key = match key {
+                    OwnedValue::Integer(i) => *i as u64,
+                    _ => unreachable!("btree tables are indexed by integers!"),
+                };
 
-            page.set_dirty();
-            self.pager.add_dirty(page.id);
+                // get page and find cell
+                let (cell_idx, page_type) = {
+                    let page = RefCell::borrow(&page_ref);
+                    if page.is_locked() {
+                        return Ok(CursorResult::IO);
+                    }
 
-            let mut page = page.contents.write().unwrap();
-            let page = page.as_mut().unwrap();
-            assert!(matches!(page.page_type(), PageType::TableLeaf));
+                    page.set_dirty();
+                    self.pager.add_dirty(page.id);
 
-            // find cell
-            (self.find_cell(page, int_key), page.page_type())
-        };
+                    let mut page = page.contents.write().unwrap();
+                    let page = page.as_mut().unwrap();
+                    assert!(matches!(page.page_type(), PageType::TableLeaf));
 
-        // TODO: if overwrite drop cell
+                    // find cell
+                    (self.find_cell(page, int_key), page.page_type())
+                };
 
-        // insert cell
+                // TODO: if overwrite drop cell
 
-        let mut cell_payload: Vec<u8> = Vec::new();
-        self.fill_cell_payload(page_type, Some(int_key), &mut cell_payload, record);
+                // insert cell
 
-        // insert
-        let overflow = {
-            let page = RefCell::borrow(&page_ref);
+                let mut cell_payload: Vec<u8> = Vec::new();
+                self.fill_cell_payload(page_type, Some(int_key), &mut cell_payload, record);
 
-            let mut page = page.contents.write().unwrap();
-            let page = page.as_mut().unwrap();
-            self.insert_into_cell(page, cell_payload.as_slice(), cell_idx);
-            page.overflow_cells.len()
-        };
+                // insert
+                let overflow = {
+                    let page = RefCell::borrow(&page_ref);
 
-        if overflow > 0 {
-            self.balance_leaf();
+                    let mut page = page.contents.write().unwrap();
+                    let page = page.as_mut().unwrap();
+                    self.insert_into_cell(page, cell_payload.as_slice(), cell_idx);
+                    page.overflow_cells.len()
+                };
+                if overflow > 0 {
+                    self.write_info.state = WriteState::BalanceStart;
```

Comment:
not super familiar with this code so asking:

we set the state to `BalanceStart` here. to get to the branch where `self.balance_leaf()` is called (below), looks like the call to `BTree::insert_to_page` (and `BTree::insert`) must happen again. what guarantees that that will happen?

Reply:
You made me sweat for a second. 

1. `InsertAsync` is called more than once on IO:
```rust
                Insn::InsertAsync {
                    cursor,
                    key_reg,
                    record_reg,
                    flag: _,
                } => {
                    let cursor = cursors.get_mut(cursor).unwrap();
                    let record = match &state.registers[*record_reg] {
                        OwnedValue::Record(r) => r,
                        _ => unreachable!("Not a record! Cannot insert a non record value."),
                    };
                    let key = &state.registers[*key_reg];
                    match cursor.insert(key, record, true)? {
                        CursorResult::Ok(_) => {
                            state.pc += 1;
                        }
                        CursorResult::IO => {
                            // If there is I/O, the instruction is restarted.
                            return Ok(StepResult::IO);
                        }
                    }
                }
```
2. It was a bug that I only call `balance_leaf` on `BalanceStart`state, it should be on every `Balance` state so I will fix that.

Reply:
Anddddddddddd 1. is somewhat wrong, because I shouldn't have returned Ok after change of state so I think you might have been referring to that. Will change it

Reply:
I ended up finding a lot more bugs so I fixed them.

Reply:
> Anddddddddddd 1. is somewhat wrong, because I shouldn't have returned Ok after change of state so I think you might have been referring to that. 

Yeah that mainly, just wasnt confident enough to suggest it was wrong. Looks more reasonable now 😆